github/install.go

```diff
@@ -0,0 +1,232 @@
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	"github.com/dlclark/regexp2"
+	"github.com/packwiz/packwiz/core"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+var GithubRegex = regexp.MustCompile(`^https?://(?:www\.)?github\.com/([^/]+/[^/]+)`)
+
+// installCmd represents the install command
+var installCmd = &cobra.Command{
+	Use:     "add [URL|slug]",
+	Short:   "Add a project from a GitHub repository URL or slug",
+	Aliases: []string{"install", "get"},
+	Args:    cobra.ArbitraryArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		pack, err := core.LoadPack()
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+
+		if len(args) == 0 || len(args[0]) == 0 {
+			fmt.Println("You must specify a GitHub repository URL.")
+			os.Exit(1)
+		}
+
+		// Try interpreting the argument as a slug, or GitHub repository URL.
+		var slug string
+		var branch string
+
+		// Regex to match potential release assets against.
+		// The default will match any asset with a name that does *not* end with:
+		// - "-api.jar"
+		// - "-dev.jar"
+		// - "-dev-preshadow.jar"
+		// - "-sources.jar"
+		// In most cases, this will only match one asset.
+		// TODO: Hopefully.
+		regex := `^.+(?<!-api|-dev|-dev-preshadow|-sources)\.jar$`
+
+		// Check if the argument is a valid GitHub repository URL; if so, extract the slug from the URL.
+		// Otherwise, interpret the argument as a slug directly.
+		matches := GithubRegex.FindStringSubmatch(args[0])
+		if len(matches) == 2 {
+			slug = matches[1]
+		} else {
+			slug = args[0]
+		}
+
+		repo, err := fetchRepo(slug)
+
+		if err != nil {
+			fmt.Printf("Failed to add project: %s\n", err)
+			os.Exit(1)
+		}
+
+		if branchFlag != "" {
+			branch = branchFlag
+		}
+		if regexFlag != "" {
+			regex = regexFlag
+		}
+
+		err = installMod(repo, branch, regex, pack)
+		if err != nil {
+			fmt.Printf("Failed to add project: %s\n", err)
+			os.Exit(1)
+		}
+	},
+}
+
+func installMod(repo Repo, branch string, regex string, pack core.Pack) error {
+	latestRelease, err := getLatestRelease(repo.FullName, branch)
+	if err != nil {
+		return fmt.Errorf("failed to get latest release: %v", err)
+	}
+
+	return installRelease(repo, latestRelease, regex, pack)
+}
+
+func getLatestRelease(slug string, branch string) (Release, error) {
+	var releases []Release
+	var release Release
+
+	resp, err := ghDefaultClient.getReleases(slug)
+	if err != nil {
+		return release, err
+	}
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return release, err
+	}
+
+	err = json.Unmarshal(body, &releases)
+	if err != nil {
+		return release, err
+	}
+
+	if branch != "" {
+		for _, r := range releases {
+			if r.TargetCommitish == branch {
+				return r, nil
+			}
+		}
+		return release, fmt.Errorf("failed to find release for branch %v", branch)
+	}
+
+	return releases[0], nil
+}
+
+func installRelease(repo Repo, release Release, regex string, pack core.Pack) error {
+	expr := regexp2.MustCompile(regex, 0)
```

Comment:
Why regexp2? I'm not entirely sure why we'd need the more powerful regex engine - would rather use Go standard libraries unless there's a specific need

Reply:
IIRC negative lookbehinds didn't work with the built-in regex.

Reply:
Ah fair. I'd probably have written it as a deny-regex instead of negative lookbehinds, most regex dialects I use don't support them.